specs/keccak256/variable-length-input.md

```diff
@@ -0,0 +1,187 @@
+
+```mermaid
+stateDiagram-v2
+    direction LR
+    [*] --> Absorb
+    Absorb --> Absorb
+    Absorb --> Finalize
+    Finalize --> Absorb
+    Finalize --> Finalize
+    Finalize --> [*]
+```
+
+#### State transition
+
+
+We leave an empty row at the top of the table for usage that disables the lookup. Use a separate selector `q_start` for this row.
+
+This region splits the input to multiple parts, each part corresponds to their `Keccak-f` permutation round.
+This is also a lookup table for the other circuits to lookup the Keccak256 input to the output.
+
+Columns:
+
+- `state_tag` either 0=Start/End/Null, 1=Absorb, 2=Finalize
+- `input_len` The length of the input.
+- `input` 136 bytes to be absorbed in this round. Padding not included yet.
+- `perm_count` Permutations we have done after the current one.
+- `acc_input` Accumulatd bytes by random linear combination (in big-endian order)
+- `output` The base-2 `state[:4]` output from this round `keccak_f`
+
+| state_tag | input_len | input | perm_count | acc_input | output |
+| --------: | --------: | ----: | ---------: | --------: | -----: |
+|         0 |         0 |     0 |          0 |         0 |      0 |
+|  Finalize |        20 |       |          1 |           |        |
+|    Absorb |       150 |       |          1 |           |        |
+|  Finalize |       150 |       |          2 |           |        |
+|  Finalize |         0 |       |          1 |           |        |
+|    Absorb |       136 |       |          1 |           |        |
+|  Finalize |       136 |     0 |          2 |           |        |
+|         0 |           |       |          0 |         0 |        |
+
+#### Checks
+
+We branch the constraints to apply by state_tag
+
+- `q_start`
+  - State transition
+    - next.state_tag in (Absorb, Finalize)
+- Absorb
+  - if input_len === 136 * (perm_count + 1) absorb a full block of 0x80...0x01
+    - next.input === 0 (since the input is the unpadded input)
+    - next.state_tag in (Absorb, Finalize)
+  - Next row validity
+    - next.acc_input === curr.acc_input * r**136
+    - next.perm_count === next.perm_count + 1
+  - State transition
+    - next.state_tag in (Absorb, Finalize)
+- Finalize
+    - This is a valid place to finalize
+        - (curr.perm_count * 136 - input_len) in 1~136
+    - Next row validity
+        - next.perm_count === 1
+    - State transition: (Absorb, Finalize, 0) all 3 states allowed
+- 0
+    - next.state_tag === 0 (The first row is also satisfied!)
+    - We can broadcast this state_tag to the `keccak_f` as a flag to disable all checks.
+
+
+#### Lookup
+
+To lookup the Keccak256 input to the output, query the following columns:
+
+- `state_tag`
+- `input_len`: FQ. This is required because input \[0, 0, 0\] and \[0, 0\] have the same RLC value but different keccak hash outputs.
+- `acc_input`: RLC
+- `output`: RLC
+
+When the lookup is needed, constrain `state_tag === 2 (Finalize)`.
```

Comment:
Wouldn't then this only trigger the last permutation round??
I'm a bit confused with that part TBH. This will basically be exposed to the end-user in the table to make an easy lookup of `RLC(input) | RLC(output) | len(input)` right? But for this to be proven correctly we need to have all of the Absorb's in the table from the first to the `Finalize` one. Isn't that the case?

Reply:
The end-user should lookup `state_tag | RLC(acc_input) | RLC(output) | input_len` and constrain `state_tag` to be `Finalize`. Note that it's `acc_input`, the full input. Not `input`, which is just a particular 17 words.